bundler/lib/bundler/checksum.rb

```diff
@@ -228,16 +223,32 @@ def to_lock(spec)
       private
 
       def register_checksum(lock_name, checksum)
-        return unless checksum
-        checksums = (store[lock_name] ||= {})
-        existing = checksums[checksum.algo]
+        existing = fetch_checksum(lock_name, checksum.algo)
 
-        if !existing
-          checksums[checksum.algo] = checksum
-        elsif existing.merge!(checksum)
-          checksum
+        if existing
+          merge_checksum(lock_name, checksum, existing)
         else
-          raise ChecksumMismatchError.new(lock_name, existing, checksum)
+          store_checksum(lock_name, checksum)
+        end
+      end
+
+      def merge_checksum(lock_name, checksum, existing)
+        existing.merge!(checksum) || raise(ChecksumMismatchError.new(lock_name, existing, checksum))
```

Comment:
Should this also be protected by the mutex?

Reply:
`merge!` just compares the digest/algo, concats the `source` onto `sources`, then runs `sources.uniq!`. Even run concurrently without global lock, I think that's still safe.

Reply:
if it's modifying any shared state, it is not safe -- remember not all Rubies have a GIL

Reply:
How can you tell which operations are safe on global state, and which aren't? This is tue code:

```
@sources.concat(other.sources).uniq!
```

This brings up another situation in here that isn't locked. I'll add a comment there.